### PR TITLE
Improves unexpected throwable handling

### PR DESCRIPTION
**Pull request by JLLeitschuh**

### CODE REVIEW

File: src/main/java/edu/wpi/grip/core/sources/CameraSource.java [hunk 1]
```diff
@@ -173,13 +173,13 @@ private synchronized void startVideo(FrameGrabber grabber) throws IOException {
                     // TODO Pass Exception to the UI.
                     System.err.println("Webcam Frame Grabber Thread crashed with uncaught exception:");
                     exception.printStackTrace();
-                    eventBus.post(new FatalErrorEvent(exception));
+                    eventBus.post(new UnexpectedThrowableEvent(exception));
                     try {
                         stopVideo();
                     } catch (TimeoutException e) {
                         System.err.println("Webcam Frame Grabber could not be stopped!");
                         e.printStackTrace();
```
Comment (ThomasJClark) [inline src/main/java/edu/wpi/grip/core/sources/CameraSource.java]: If we're gonna start actually handling errors, don't print the stack trace here.  We'll have something listening for stack traces to log

File: src/main/java/edu/wpi/grip/ui/Main.java [hunk 2]
```diff
@@ -43,18 +42,22 @@ public void start(Stage stage) {
         stage.show();
     }
 
+    private void triggerUnexpectedThrowableEvent(UnexpectedThrowableEvent event) {
```
Comment (ThomasJClark) [inline src/main/java/edu/wpi/grip/ui/Main.java]: Why is this wrapped in a method?

File: src/main/java/edu/wpi/grip/ui/Main.java [hunk 3]
```diff
@@ -62,5 +65,11 @@ public final void onFatalErrorEvent(FatalErrorEvent error) {
                 }
             }
         });
+
+        if(event.isFatal()) {
+            System.err.println("Original fatal exception");
+            event.getThrowable().printStackTrace();
```
Comment (ThomasJClark) [inline src/main/java/edu/wpi/grip/ui/Main.java]: We already print the stack trace [here](https://github.com/WPIRoboticsProjects/GRIP/pull/179/files#diff-d5c3d322c479e7a4fc933fa793c49bcdR52), right?

File: src/main/java/edu/wpi/grip/ui/pipeline/AddSourceView.java [hunk 4]
```diff
@@ -52,7 +52,7 @@ public AddSourceView(EventBus eventBus) {
                     eventBus.post(new SourceAddedEvent(new ImageFileSource(eventBus, file)));
                 } catch (IOException e) {
                     e.printStackTrace();
```
Comment (ThomasJClark) [inline src/main/java/edu/wpi/grip/ui/pipeline/AddSourceView.java]: Remove this `printStackTrace()`s, since we have a handler that prints them

File: src/main/java/edu/wpi/grip/core/events/UnexpectedThrowableEvent.java [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+public final class UnexpectedThrowableEvent {
+    private final Throwable throwable;
+    private final boolean fatal;
+
+    /**
+     * @param throwable The throwable that was caught.
+     * @param fatal     True if this cause the application to quit forcibly.
+     *                  If the throwable is an {@link Error} than this is automatically true. Defaults to false.
+     */
+    public UnexpectedThrowableEvent(Throwable throwable, boolean fatal) {
```
Comment (ThomasJClark) [inline src/main/java/edu/wpi/grip/core/events/UnexpectedThrowableEvent.java]: I'm assuming `fatal` is true if we send this from a `Thread.UncaughtExceptionHandler`?  In that case, you should add handling to any threads we create

File: src/main/java/edu/wpi/grip/core/events/UnexpectedThrowableEvent.java [hunk 6]
```diff
@@ -0,0 +1,37 @@
+package edu.wpi.grip.core.events;
+
+/**
+ * Event should be thrown when Unexpected Throwable ends up in an {@link Thread#uncaughtExceptionHandler}.
```
Comment (ThomasJClark) [inline src/main/java/edu/wpi/grip/core/events/UnexpectedThrowableEvent.java]: Based on your usage of this event, this this isn't true

File: src/main/java/edu/wpi/grip/ui/Main.java [hunk 7]
```diff
@@ -31,8 +31,7 @@ public void start(Stage stage) {
          * Any exceptions thrown by the UI will be caught here and an exception dialog will be displayed
          */
         Thread.currentThread().setUncaughtExceptionHandler((thread, throwable) -> {
-            this.eventBus.post(new FatalErrorEvent(throwable));
-
+            this.eventBus.post(new UnexpectedThrowableEvent(throwable));
```
Comment (ThomasJClark) [inline src/main/java/edu/wpi/grip/ui/Main.java]: This should be fatal

File: src/main/java/edu/wpi/grip/core/events/UnexpectedThrowableEvent.java (hunk 5 above)
Comment (JLLeitschuh) [inline src/main/java/edu/wpi/grip/core/events/UnexpectedThrowableEvent.java]: I plan to. In a different PR. I'm trying to go for small pull requests that are easy to review and merge.

File: src/main/java/edu/wpi/grip/core/events/UnexpectedThrowableEvent.java (hunk 6 above)
Comment (JLLeitschuh) [inline src/main/java/edu/wpi/grip/core/events/UnexpectedThrowableEvent.java]: I was going to fix that in a separate PR

File: src/main/java/edu/wpi/grip/ui/Main.java (hunk 7 above)
Comment (JLLeitschuh) [inline src/main/java/edu/wpi/grip/ui/Main.java]: Why?
JavaFX hits null pointers all the time. I'm working on fixing that. Some of these things can be easily resolved by simply closing and reopening a preview for example.

File: src/main/java/edu/wpi/grip/ui/pipeline/AddSourceView.java (hunk 4 above)
Comment (JLLeitschuh) [inline src/main/java/edu/wpi/grip/ui/pipeline/AddSourceView.java]: Not what this PR is about.

File: src/main/java/edu/wpi/grip/ui/Main.java (hunk 3 above)
Comment (JLLeitschuh) [inline src/main/java/edu/wpi/grip/ui/Main.java]: I want to make it entirely clear what the original stack trace is just in case more come later.

File: src/main/java/edu/wpi/grip/core/sources/CameraSource.java (hunk 1 above)
Comment (JLLeitschuh) [inline src/main/java/edu/wpi/grip/core/sources/CameraSource.java]: Events should not be used to log. Logging is a separate thing that adds overhead.

File: src/main/java/edu/wpi/grip/core/events/UnexpectedThrowableEvent.java (hunk 6 above)
Comment (ThomasJClark) [inline src/main/java/edu/wpi/grip/core/events/UnexpectedThrowableEvent.java]: Fix the comment or change it so the comment is true?

File: src/main/java/edu/wpi/grip/ui/pipeline/AddSourceView.java (hunk 4 above)
Comment (ThomasJClark) [inline src/main/java/edu/wpi/grip/ui/pipeline/AddSourceView.java]: I would argue that fixing the double or triple stack trace printed would definitely "improve unexpected throwable handling"

File: src/main/java/edu/wpi/grip/ui/Main.java (hunk 7 above)
Comment (ThomasJClark) [inline src/main/java/edu/wpi/grip/ui/Main.java]: I've never seen this.  Can you open an issue for it?

File: src/main/java/edu/wpi/grip/ui/Main.java (hunk 3 above)
Comment (ThomasJClark) [inline src/main/java/edu/wpi/grip/ui/Main.java]: What?  This just makes it print the same stack trace twice

File: src/main/java/edu/wpi/grip/core/sources/CameraSource.java (hunk 1 above)
Comment (ThomasJClark) [inline src/main/java/edu/wpi/grip/core/sources/CameraSource.java]: :disappointed: 

https://github.com/WPIRoboticsProjects/GRIP/pull/179/files#diff-d5c3d322c479e7a4fc933fa793c49bcdR52

Comment (JLLeitschuh) [inline src/main/java/edu/wpi/grip/core/sources/CameraSource.java]: Logging things that are expected parts of the application. We should definitely use events for reporting errors to the UI

File: src/main/java/edu/wpi/grip/ui/Main.java (hunk 2 above)
Comment (JLLeitschuh) [inline src/main/java/edu/wpi/grip/ui/Main.java]: Because I can't access the `eventBus` from within a lambda being created inside of a constructor for a field that hasn't been assigned yet.

File: src/main/java/edu/wpi/grip/ui/Main.java (hunk 3 above)
Comment (JLLeitschuh) [inline src/main/java/edu/wpi/grip/ui/Main.java]: If your console doesn't have infinite scroll on and another exception is thrown from anywhere inside of `PlatformImpl.runAndWait(...` (IE sun's code not our code) then it will be lost.

File: src/main/java/edu/wpi/grip/ui/pipeline/AddSourceView.java (hunk 4 above)
Comment (JLLeitschuh) [inline src/main/java/edu/wpi/grip/ui/pipeline/AddSourceView.java]: Fine.

File: src/main/java/edu/wpi/grip/core/events/UnexpectedThrowableEvent.java (hunk 6 above)
Comment (JLLeitschuh) [inline src/main/java/edu/wpi/grip/core/events/UnexpectedThrowableEvent.java]: Nope, as soon as I'm done with this I'm going to be replacing its invalid usages with something else. In the meantime it stays where it is to handle exceptions we get.

File: src/main/java/edu/wpi/grip/ui/Main.java (hunk 7 above)
Comment (JLLeitschuh) [inline src/main/java/edu/wpi/grip/ui/Main.java]: Its open: #155

File: src/main/java/edu/wpi/grip/core/events/UnexpectedThrowableEvent.java (hunk 6 above)
Comment (ThomasJClark) [inline src/main/java/edu/wpi/grip/core/events/UnexpectedThrowableEvent.java]: Nope as in neither?

File: src/main/java/edu/wpi/grip/core/sources/CameraSource.java (hunk 1 above)
Comment (ThomasJClark) [inline src/main/java/edu/wpi/grip/core/sources/CameraSource.java]: I'm having difficulty interpreting this comment.  Could you clarify/rephrase that?

Comment (ThomasJClark) [inline src/main/java/edu/wpi/grip/core/sources/CameraSource.java]: - You're logging an exception then sending the exception in an event
- The event handler also logs the exception
- You said "Events should not be used to log", but that's what you're doing

File: src/main/java/edu/wpi/grip/core/events/UnexpectedThrowableEvent.java (hunk 6 above)
Comment (JLLeitschuh) [inline src/main/java/edu/wpi/grip/core/events/UnexpectedThrowableEvent.java]: Yes. I'm going to remove it. Just not right now. That is coming in another PR.

File: src/main/java/edu/wpi/grip/core/sources/CameraSource.java (hunk 1 above)
Comment (JLLeitschuh) [inline src/main/java/edu/wpi/grip/core/sources/CameraSource.java]: I'll remove this stack trace.
In the future however, logging should be done in place where the exception occurs.

File: ui/src/main/java/edu/wpi/grip/ui/ExceptionAlert.java [hunk 8]
```diff
@@ -62,17 +64,20 @@
      *                  the issue website.
      * @see <a href="http://code.makery.ch/blog/javafx-dialogs-official/">Inspiration</a>
      */
-    public ExceptionAlert(final Parent root, final Throwable throwable, final HostServices services) {
+    public ExceptionAlert(final Parent root, final Throwable throwable, Optional<String> message, boolean isFatal, final HostServices services) {
```
Comment (ThomasJClark) [inline ui/src/main/java/edu/wpi/grip/ui/ExceptionAlert.java]: `Optional<String> message` should just be an optional parameter

File: ui/src/main/java/edu/wpi/grip/ui/ExceptionAlert.java [hunk 9]
```diff
@@ -203,10 +199,9 @@ private String generateSystemInfoMessage() {
      * @return The fully constructed issue text.
      */
     private String issueText() {
-        return new StringBuilder(ISSUE_PROMPT_QUESTION)
-                .append("\n\n\n\n")
-                .append(systemInfoMessage)
-                .append(exceptionMessage).toString();
+        final StringBuilder builder = new StringBuilder(ISSUE_PROMPT_QUESTION).append("\n\n\n\n")
```
Comment (ThomasJClark) [inline ui/src/main/java/edu/wpi/grip/ui/ExceptionAlert.java]: Use string concatenation

Comment (ThomasJClark) [inline ui/src/main/java/edu/wpi/grip/ui/ExceptionAlert.java]: is this event used?

File: ui/src/main/java/edu/wpi/grip/ui/ExceptionAlert.java [hunk 10]
```diff
@@ -203,10 +199,12 @@ private String generateSystemInfoMessage() {
      * @return The fully constructed issue text.
      */
     private String issueText() {
-        return new StringBuilder(ISSUE_PROMPT_QUESTION)
-                .append("\n\n\n\n")
-                .append(systemInfoMessage)
-                .append(exceptionMessage).toString();
+        return ISSUE_PROMPT_QUESTION
+                + "\n\n\n\n"
+                + getAdditionalInfoMessage()
```
Comment (JLLeitschuh) [inline ui/src/main/java/edu/wpi/grip/ui/ExceptionAlert.java]: Undo this change....
